backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/ColumnarCachedBatchSerializer.scala

```diff
@@ -0,0 +1,303 @@
+      it =>
+        RowToVeloxColumnarExec.toColumnarBatchIterator(
+          it,
+          localSchema,
+          numInputRows,
+          numOutputBatches,
+          convertTime,
+          numRows
+        )
+    }
+    convertColumnarBatchToCachedBatch(rddColumnarBatch, schema, storageLevel, conf)
+  }
+
+  override def convertCachedBatchToInternalRow(
+      input: RDD[CachedBatch],
+      cacheAttributes: Seq[Attribute],
+      selectedAttributes: Seq[Attribute],
+      conf: SQLConf): RDD[InternalRow] = {
+    if (!validateSchema(cacheAttributes)) {
+      // if we do not support this schema that means we are using row-based serializer,
+      // see `convertInternalRowToCachedBatch`, so fallback to vanilla Spark serializer
+      return rowBasedCachedBatchSerializer.convertCachedBatchToInternalRow(
+        input,
+        cacheAttributes,
+        selectedAttributes,
+        conf)
+    }
+
+    // note, these metrics are unused but just make `VeloxColumnarToRowExec` happy
+    val metrics = BackendsApiManager.getMetricsApiInstance.genColumnarToRowMetrics(
+      SparkSession.getActiveSession.orNull.sparkContext)
+    val numOutputRows = metrics("numOutputRows")
+    val numInputBatches = metrics("numInputBatches")
+    val convertTime = metrics("convertTime")
+    val rddColumnarBatch =
+      convertCachedBatchToColumnarBatch(input, cacheAttributes, selectedAttributes, conf)
+    rddColumnarBatch.mapPartitions {
+      it =>
+        VeloxColumnarToRowExec.toRowIterator(
+          it,
+          selectedAttributes,
+          numOutputRows,
+          numInputBatches,
+          convertTime
+        )
+    }
+  }
+
+  override def convertColumnarBatchToCachedBatch(
+      input: RDD[ColumnarBatch],
+      schema: Seq[Attribute],
+      storageLevel: StorageLevel,
+      conf: SQLConf): RDD[CachedBatch] = {
+    val timezoneId = SQLConf.get.sessionLocalTimeZone
+    val localSchema = toStructType(schema)
+    input.mapPartitions {
+      it =>
+        val allocator = ArrowBufferAllocators.contextInstance()
+        val arrowSchema = SparkArrowUtil.toArrowSchema(localSchema, timezoneId)
+
+        new Iterator[CachedBatch] {
+          override def hasNext: Boolean = it.hasNext
+
+          override def next(): CachedBatch = {
+            val batch = it.next()
+            val vsr = ColumnarBatches.loadArrowVectorSchemaRoot(allocator, batch, arrowSchema)
+            try {
+              val vectorUnloader = new VectorUnloaderWithField(vsr)
+              val numRows = vsr.getRowCount
+              val (sizeInBytes, fieldRecordBatch) = vectorUnloader.unloadToFieldRecordBatch
+              CachedColumnarBatch(
+                numRows,
+                sizeInBytes,
+                fieldRecordBatch.fieldBuffersArr,
+                fieldRecordBatch.fieldNullCountsArr,
+                fieldRecordBatch.fieldNumRowsArr
+              )
+            } finally {
+              vsr.close()
+            }
```

Comment:
When will the input batch get closed?

Reply:
inside `ColumnarBatches.loadArrowVectorSchemaRoot`, I have added comment for that method